Fix start_date and hours validation for meetings

oliverguenther opened the pull request:
This fixes a number of issues:
1. When creating meetings, invalid dates or hour fields were not marked
   as errors since the value is a composite of both values.
   This is fixed by creating explicit setters and only converting to the
   composite value just before saving.
2. The date error message was not specifying what format is required.
3. The default start_date value was written when the date was erroneous

https://community.openproject.com/work_packages/20449/activity
https://community.openproject.com/work_packages/23491

oliverguenther committed:
Fix start_date and hours validation for meetings

This fixes a number of issues:

1. When creating meetings, invalid dates or hour fields were not marked
as errors since the value is a composite of both values.
This is fixed by creating explicit setters and only converting to the
composite value just before saving.

2. The date error message was not specifying what format is required.

3. The default start_date value was written when the date was erroneous

oliverguenther commented:
@ulferts , @machisuji LFR.

This is the solution I went with. This has the side-effect that `start_time` isn't re-set until the model is saved - I could add that but we don't require it and it adds some more complexity to the model.

All other solutions including `#attribute_will_change!` were not satisfactory since they require some kind of memory for the virtual attributes.

ulferts commented:
I think the approach is reasonable. 

I'd like the following little changes:

![image](https://cloud.githubusercontent.com/assets/617519/16264038/7c8abe20-3877-11e6-8509-05193fbad17b.png)

1) Keep the values even after a faulty input 
2) Different string for "Start time hour is invalid". e.g. "Start time is invalid. Required format: HH:MM"
3) As whatever value one might have stored in `start_time` gets overwritten in a before_save hook, I would find it nice as a developer if I was notified of such a behaviour. e.g. prevent setting a value from outside the class (will be hard because of the different setters available) or throw an exception in the before_save if the value is already changed.

oliverguenther committed:
Fix sort_by since start_date is now a string

oliverguenther commented:
@ulferts  I've now resorted to the `attribute_will_change!` implementation and by that addressed all your remarks except 3.
1. `start_date` and `start_time_hour` now simply accept any input, and only before save is the `start_time` object written based on the dirty check and after the input was validated.
2. Done
3. We cannot raise or override the `start_time=`  setter due to the way journals recreate initial data using attributes from models. For meetings, this causes `start_time` to be assigned through `assign_attributes`, which internally call the setter.

I believe it's okay to leave `start_time` available to developers. I simply check now whether `start_date` or `start_time_hour` was changed and use that then. Otherwise, start_time is checked itself , which rails sets to nil when an invalid date was passed.

For start_date, the faulty output is correctly returned:
![bildschirmfoto 2016-06-22 um 14 01 38](https://cloud.githubusercontent.com/assets/459462/16266003/dd7fb3e2-3882-11e6-8914-5d25083f27a6.png)

However due to `type="time"`, the faulty string is replaced by `--:--`, which should be fine.

![bildschirmfoto 2016-06-22 um 14 01 53](https://cloud.githubusercontent.com/assets/459462/16266005/df593ae4-3882-11e6-9c70-6b7660ca735a.png)

ulferts commented:
Please add an l10n string for :start_time_hour on meetings so that it is localized properly, e.g. "Start time"

oliverguenther committed:
Add start time locale

oliverguenther commented:
@ulferts Done

ulferts commented:
:+1:
